packages/amplify-migration-e2e/src/templategen.ts

```diff
@@ -0,0 +1,143 @@
+
+export type RefactorCategory = 'auth' | 'storage';
+
+const RETRY_CONFIG: RetrySettings = {
+  times: 50,
+  delayMS: 1000, // 1 second
+  timeoutMS: 1000 * 60 * 5, // 5 minutes
+  stopOnError: true,
+};
+
+const STATUS_COMPLETE = 'COMPLETE';
+const STATUS_IN_PROGRESS = 'IN_PROGRESS';
+const STATUS_FAILED = 'FAILED';
+
+export function runTemplategenCommand(cwd: string, gen1StackName: string, gen2StackName: string) {
+  const parentDir = path.resolve(cwd, '..');
+  const processResult = execa.sync(
+    getNpxPath(),
+    ['--prefix', parentDir, '@aws-amplify/migrate', 'to-gen-2', 'generate-templates', '--from', gen1StackName, '--to', gen2StackName],
+    {
+      cwd,
+      env: { ...process.env, npm_config_user_agent: 'npm' },
+      encoding: 'utf-8',
+    },
+  );
+
+  if (processResult.exitCode !== 0) {
+    throw new Error(`Templategen command exit code: ${processResult.exitCode}, message: ${processResult.stderr}`);
+  }
+}
+
+function uncommentS3BucketLineFromBackendFile(projRoot: string) {
+  const backendFilePath = path.join(projRoot, 'amplify', 'backend.ts');
+  const backendFileContent = fs.readFileSync(backendFilePath, 'utf8');
+  const regex = /^\s*\/\/\s*(s3Bucket\.bucketName)/m;
+  const updatedBackendFileContent = backendFileContent.replace(regex, '$1');
+  fs.writeFileSync(backendFilePath, updatedBackendFileContent);
+}
+
+async function executeCommand(command: string, cwd?: string) {
+  cwd = cwd ?? process.cwd();
+  const processResult = execa.sync(command, {
+    cwd,
+    env: { ...process.env, npm_config_user_agent: 'npm' },
+    encoding: 'utf-8',
+    shell: true,
+  });
+  if (processResult.exitCode === 0) {
+    return processResult.stdout;
+  } else {
+    throw new Error(`Command exit code: ${processResult.exitCode}, message: ${processResult.stderr}`);
+  }
+}
+
+async function executeCreateStackRefactorCallCommand(command: string, cwd: string) {
+  const processResult = JSON.parse(await executeCommand(command, cwd));
+  const stackRefactorId = processResult.StackRefactorId;
+  return stackRefactorId;
+}
+
+async function executeStep1(cwd: string, commands: string[]) {
+  await executeCommand(commands[0], cwd);
+  await retry(
+    () => assertStepCompletion(commands[1]),
+    (status) => status.includes(STATUS_COMPLETE) && !status.includes(STATUS_IN_PROGRESS),
+    RETRY_CONFIG,
+    (status) => status.includes(STATUS_FAILED),
+  );
+}
+
+async function executeStep2(cwd: string, commands: string[]) {
+  await executeCommand(commands[0], cwd);
+  await retry(
+    () => assertStepCompletion(commands[1]),
+    (status) => status.includes(STATUS_COMPLETE) && !status.includes(STATUS_IN_PROGRESS),
+    RETRY_CONFIG,
+    (status) => status.includes(STATUS_FAILED),
+  );
+}
+
+async function executeStep3(cwd: string, commands: string[], bucketName: string) {
+  toggleEnvVariable('BUCKET_NAME', 'SET', bucketName);
+  await executeCommand(commands[1], cwd);
```

Comment:
eferencing the comment from the previous PR: https://github.com/aws-amplify/amplify-cli/pull/13984#discussion_r1809356785